homeassistant/components/ring/binary_sensor.py

```diff
@@ -53,70 +69,88 @@ async def async_setup_entry(
+        and async_check_create_deprecated(
+            hass,
+            Platform.BINARY_SENSOR,
+            f"{device.id}-{description.key}",
+            description,
+        )
+    )
 
 
 class RingBinarySensor(
-    RingBaseEntity[RingNotificationsCoordinator], BinarySensorEntity
+    RingBaseEntity[RingListenCoordinator, RingDeviceT], BinarySensorEntity
 ):
     """A binary sensor implementation for Ring device."""
 
     _active_alert: RingEvent | None = None
-    entity_description: RingBinarySensorEntityDescription
+    RingBinarySensorEntityDescription[RingDeviceT]
 
     def __init__(
         self,
-        ring: Ring,
-        device: RingGeneric,
-        coordinator: RingNotificationsCoordinator,
-        description: RingBinarySensorEntityDescription,
+        device: RingDeviceT,
+        coordinator: RingListenCoordinator,
+        description: RingBinarySensorEntityDescription[RingDeviceT],
     ) -> None:
-        """Initialize a sensor for Ring device."""
+        """Initialize a binary sensor for Ring device."""
         super().__init__(
             device,
             coordinator,
         )
         self.entity_description = description
-        self._ring = ring
         self._attr_unique_id = f"{device.id}-{description.key}"
-        self._update_alert()
+        self._attr_is_on = False
+        self._active_alert: RingEvent | None = None
+        self._cancel_callback: TimerHandle | None = None
 
     @callback
-    def _handle_coordinator_update(self, _: Any = None) -> None:
-        """Call update method."""
-        self._update_alert()
-        super()._handle_coordinator_update()
+    def _async_handle_event(self, alert: RingEvent) -> None:
+        """Handle the event."""
+        self._attr_is_on = True
+        self._active_alert = alert
+        loop = self.hass.loop
+        when = loop.time() + alert.expires_in
+        if self._cancel_callback:
+            self._cancel_callback.cancel()
+        self._cancel_callback = loop.call_at(when, self._async_cancel_event)
 
     @callback
-    def _update_alert(self) -> None:
-        """Update active alert."""
-        self._active_alert = next(
-            (
-                alert
-                for alert in self._ring.active_alerts()
-                if alert["kind"] == self.entity_description.key
-                and alert["doorbot_id"] == self._device.id
-            ),
-            None,
+    def _async_cancel_event(self) -> None:
+        """Clear the event."""
+        self._cancel_callback = None
+        self._attr_is_on = False
+        self._active_alert = None
+        self.async_write_ha_state()
+
+    def _get_coordinator_alert(self) -> RingEvent | None:
+        alerts = (
+            alert
+            for alert in self.coordinator.ring_api.active_alerts()
+            if alert.kind == self.entity_description.key
+            and alert.doorbot_id == self._device.device_api_id
         )
+        return next(alerts, None)
```

Comment:
It would be nicer if the alters were indexed so we didn't have to iterate them every time

Reply:
Done